### ISSUE TITLE
Volume Label does not update correctly

### ISSUE THREAD

**Issue by cnitschkowski** [billziss-gh/winfsp#236]
## Bug Report

When I call FileSystemBase.SetVolumeLabel the first time, it sets the volume label accordingly in most cases. When I update the volume label again calling FileSystemBase.SetVolumeLabel while the filesystem is mounted, it does not change but stays at the label I assigned to it first.

### How to Reproduce

1. Implement filesystem using .NET.
2. Call FileSystemHost.FileSystem().SetVolumeLabel("A", out VolumeInfo info);
3. Check the label on cmd using "dir"
4. Call FileSystemHost.FileSystem().SetVolumeLabel("B", out VolumeInfo info);
5. Check the label on cmd using "dir"

### Behaviors

I'd expect the volume label to be "A" in step 3 and "B" in step 5.
After step 2 it changes to "A" and stays like that until the filesystem host process is terminated no matter how often SetVolumeLabel gets called.

### Environment

- OS version and build: 10 1803 17134.590
- WinFsp version and build: 1.4.18337

**Comment by cnitschkowski** [comment]
I upgraded to WinFsp version 1.4.19049 and the problem is still there.

**Comment by billziss-gh** [comment]
> 1. Implement filesystem using .NET.
> 2. Call FileSystemHost.FileSystem().SetVolumeLabel("A", out VolumeInfo info);
> 3. Check the label on cmd using "dir"
> 4. Call FileSystemHost.FileSystem().SetVolumeLabel("B", out VolumeInfo info);
> 5. Check the label on cmd using "dir"

**Summary**

`SetVolumeLabel` is a method that is called by the OS to inform the file system that it needs to change the volume label. In general it makes little sense to call `SetVolumeLabel` from your code and if you do so the file system may update the volume label in its internal data structures, but the OS and WinFsp FSD has no direct knowledge of the change.

**Explanation**

When you change the volume label with a Win32 API like [`SetVolumeLabelW`](https://docs.microsoft.com/en-us/windows/desktop/api/winbase/nf-winbase-setvolumelabelw) what happens is that the OS constructs an IRP (I/O Request Packet) that describes the volume label change and then forwards it to the WinFsp FSD. The WinFsp FSD in turn forwards the IRP to the user mode file system, which eventually results in a call to `SetVolumeLabel` on your file system. The file system updates its internal data structures to reflect the volume change and returns the updated volume information, which is used by the WinFsp FSD to complete the original IRP.

Of importance to resolving your issue is that the WinFsp FSD caches the volume information just received for a timeout that is controlled by the `FileInfoTimeout` property. So if the Win32 API [`GetVolumeInformationW`](https://docs.microsoft.com/en-us/windows/desktop/api/fileapi/nf-fileapi-getvolumeinformationw) is issued before the timeout expires, the resulting IRP will be fully handled by the WinFsp FSD and your file system's `GetVolumeInfo` will not be called (this is done for optimization purposes).

You can now see that if you change the volume label inside the file system (via calling directly the file system's `SetVolumeLabel` or other method) it may not affect the volume label reported to the OS, because the WinFsp FSD caches this information.

So my question is: what do you set the `FileInfoTimeout` as?

**Comment by cnitschkowski** [comment]
Hello Bill,
it took me a month to check this because of a(nother) shift in priorieties.
The FileInfoTimeout was set to uint.MaxValue.
After changing it to 1000 (1 second), it works as expected.
You told me in a mail that you consider exposing FileInfoTimeout in "FileInfoTimeout", "DirInfoTimeout" and "VolumeInfoTimeout" to the .NET API. Are you still planning on doing so? That'd be great.

I'll close the issue as my problem doesn't exist anymore with proper timeouts set.

**Comment by billziss-gh** [comment]
@cnitschkowski thanks. I will open a new issue to track this feature request.
